bin/cloudtrail_handler.py

```diff
@@ -39,36 +39,68 @@ def __init__(self, config, s3, query_dao):
         self.query_dao = query_dao
 
     def process_log(self, event):
-        s3_record = event['Records'][0]['s3']
-        bucket = s3_record['bucket']['name']
-        key = s3_record['object']['key']
+        """Process the CloudTrail log event for Athena queries."""
+        s3_record = event["Records"][0]["s3"]
+        bucket = s3_record["bucket"]["name"]
+        key = s3_record["object"]["key"]
         object = self.s3.Object(bucket, key).get()
-        stream = BytesIO(object['Body'].read())
-        with gzip.GzipFile(None, 'rb', fileobj=stream) as file:
+        stream = BytesIO(object["Body"].read())
+        with gzip.GzipFile(None, "rb", fileobj=stream) as file:
             try:
                 data = json.load(file)
             except json.JSONDecodeError:
-                logger.warning(f'Not a valid cloudtrail json file {bucket}/{key}', exc_info=True)
+                logger.warning(f"Not a valid cloudtrail json file {bucket}/{key}", exc_info=True)
                 return
-            for record in data.get('Records', []):
-                if record.get('eventName') == 'StartQueryExecution':
+            for record in data.get("Records", []):
+                if record.get("eventName") == "StartQueryExecution":
                     self.process_query_record(record)
 
+    def extract_user_from_arn(self, arn):
+        """Extract the user identifier from an ARN."""
+        arn_parts = arn.split("/")
+        if len(arn_parts) >= 3:
+            email = arn_parts[-1]
+            if "@" in email:
+                return email.split("@")[0]
+            else:
+                return email
+        return "Unknown-Email"
+
+    def infer_executing_user(self, user_identity):
+        """Determine the user who executed the query based on identity info."""
+        if not user_identity:
+            return "Unknown"
+
+        identity_type = user_identity.get("type")
+        if identity_type == "IAMUser":
+            return user_identity.get("userName", "Unknown-IAMUser")
+        if identity_type == "AssumedRole" and "arn" in user_identity:
+            return self.extract_user_from_arn(user_identity["arn"])
+        return "API"
```

Comment:
Small change but I'd suggest to rename it to "Unknown" too, or "Failed to parse". With this we will know if it was Lambda parsing issue or just a query done by SSO:D

Reply:
I think it was there signify actual API-made queries - when it has an identity but it's neither IAM nor arn.